### ISSUE TITLE
Add Decoder::skip method

### ISSUE THREAD

**Issue by tailhook** [twittner/cbor-codec#3]
Sometimes when reading complex structure, I want to skip some values,  to achieve the following tasks as an example:

1. Skipping unknown keys when reading structure, for forward compatibility
2. Reading only a few keys from a larger structure
3. Filter list of serialized objects by some criteria

All this needs `decoder.skip()` method which reads *any* object (i.e. arbitrary complex structure) and discards it while reading (comparing to GenericReader which will allocate/hold structure in memory until it returns to user code).

What do you think?

**Comment by twittner** [comment]
Sounds good. I'll add it soon.

**Comment by twittner** [comment]
I have pushed an initial implementation to a [branch](https://github.com/twittner/cbor-codec/compare/skip). What do you think?

**Comment by tailhook** [comment]
Looks good, except `skip_value` thing which looks weird. Have you considered to implement `skip()` method only on readers which also implement `Seek` trait?

**Comment by twittner** [comment]
> Looks good, except `skip_value` thing which looks weird.

Could you elaborate? `skip_value` is private and exists primarily to limit recursion depth and handle `break` (indefinite collections).

> Have you considered to implement `skip()` method only on readers which also implement `Seek` trait?

Yes, but I think it would be overly restrictive as it prevents using this decoder with streaming `Read` implementations.

**Comment by tailhook** [comment]
> Could you elaborate? skip_value is private and exists primarily to limit recursion depth and handle break (indefinite collections).

Reading by 512 bytes chunks, and copying bytes to the stack where you just need a counter increment. I.e. most of the time I use cbor decoder on a buffer (`Cursor<Vec<u8>>`), and even on files I think that `seek()` can't be more expensive than a `read()` :)

>> Have you considered to implement skip() method only on readers which also implement Seek trait?
> Yes, but I think it would be overly restrictive as it prevents using this decoder with streaming Read implementations.

Partially true, it restricts using `skip()` method **only**. I don't believe that decoding directly from a network stream is very common case. Usually packet is read fully into memory before starting to decode.

So I'd rather limit to Seek trait, until there is somebody with an actual use case of skip method for un-Seek'able stream. Having concrete use case we can see if it's even reasonable to overcome the limitation by copying bytes or it will kill performance too much (i.e. for reading from socket, buffer of 512 bytes is usually too small).

**Comment by twittner** [comment]
> Reading by 512 bytes chunks, and copying bytes to the stack where you just need a counter increment.

How would you do this without `std::io::Seek`?

> I don't believe that decoding directly from a network stream is very common case. Usually packet is read fully into memory before starting to decode.

Maybe, but a socket-like type would still not be able to implement `std::io::Seek` (cf. `std::net::TcpStream`).

> for reading from socket, buffer of 512 bytes is usually too small

True, but I would expect clients to use `std::io::BufRead` in this case.

What I would ideally like to have is a `Skip` trait:

```rust
pub trait Skip {
    fn skip(n: usize) -> Result<()>;
}
```

The various implementations could optimise this case as much as possible and it would allow an implementation for `TcpStream` and other types which can go "forward" only.

**Comment by tailhook** [comment]
> What I would ideally like to have is a Skip trait

I'm all for that. But there are too many questions: should it be in this crate or separate? Which types to support out of the box? Why would users implement the trait for their own types? etc.

I would use `Seek` trait for now. Then if it will be needed by anyone it's easy to create `Skip` trait that is implemented for all types having `Seek`, which should be backwards-compatible AFAIU.

**Comment by twittner** [comment]
I would rather define a `Skip` trait now to give users the option of defining an implementation for their own types if necessary and provide an implementation for `Seek` only. If a general purpose `Skip` trait emerges somewhere we can migrate to it, but I would not want to constrain the API if possible.

The `skip` branch has been updated accordingly. Let me know what you think.

**Comment by tailhook** [comment]
Looks Ok. Except you may need to declare Skip trait for all the types in `std` which don't have `Seek` trait defined. Because it's impossible to do in any crate other that the one where trait is declared (it's not needed for my current use case though, so you may leave this on TODO list)

**Comment by twittner** [comment]
Since we considered only supporting `skip` for types which implement `Seek` I thought it is fair to only provide an implementation of `Skip` for these types. If someone wants to write an implementation for other types from `std` (e.g. `TcpStream`) a newtype could be used and we can add more implementations as needed. If that's acceptable I would merge the branch and make a new release, alright?

**Comment by tailhook** [comment]
Yes. It's acceptable for me now.

**Comment by twittner** [comment]
Released as version 0.2.0 to [crates.io](https://crates.io/crates/cbor-codec). Thanks again for your insightful remarks.
